From e090c599ef76f4eb696c38b3a210d32e65c57fdb Mon Sep 17 00:00:00 2001
From: Michal Hrusecky <michal.hrusecky@turris.com>
Date: Thu, 6 May 2021 22:37:56 +0200
Subject: [PATCH] alert_luajit: Make destination IP address available

Make sure we have destination IP address available at hand in lua
alerting modules.
---
 src/loggers/alert_luajit.cc    | 8 ++++++++
 src/managers/lua_plugin_defs.h | 2 +-
 src/protocols/ipv6.h           | 1 +
 3 files changed, 10 insertions(+), 1 deletion(-)

diff --git a/src/loggers/alert_luajit.cc b/src/loggers/alert_luajit.cc
index a6fb19d96..6e1797056 100644
--- a/src/loggers/alert_luajit.cc
+++ b/src/loggers/alert_luajit.cc
@@ -22,6 +22,7 @@
 #endif
 
 #include "protocols/eth.h"
+#include "protocols/ipv6.h"
 #include "detection/ips_context.h"
 #include "detection/signature.h"
 #include "events/event.h"
@@ -82,6 +83,7 @@ SO_PUBLIC const SnortPacket* get_packet()
     lua_packet.num = packet->context->packet_number;
     lua_packet.sp = packet->ptrs.sp;
     lua_packet.dp = packet->ptrs.dp;
+    lua_packet.dst_addr = "";
 
     if ( !(packet->proto_bits & PROTO_BIT__ETH) ) {
         lua_packet.ether_dst = lua_packet.ether_src = "";
@@ -112,6 +114,12 @@ SO_PUBLIC const SnortPacket* get_packet()
         lua_packet.ether_dst = "";
     }
 
+    if ( packet->has_ip() or packet->is_data() ) {
+        static THREAD_LOCAL char ip[ip::IP6_MAX_STR_LEN];
+        packet->ptrs.ip_api.get_dst()->ntop(ip, ip::IP6_MAX_STR_LEN);
+        lua_packet.dst_addr = ip;
+    }
+
     return &lua_packet;
 }
 
diff --git a/src/managers/lua_plugin_defs.h b/src/managers/lua_plugin_defs.h
index 5fce3682e..7de831f81 100644
--- a/src/managers/lua_plugin_defs.h
+++ b/src/managers/lua_plugin_defs.h
@@ -50,13 +50,13 @@ const struct SnortEvent* get_event();
 
 struct SnortPacket
 {
-    // FIXIT-L add ip addrs and other useful foo to lua packet
     const char* type;
     uint64_t num;
     unsigned sp;
     unsigned dp;
     const char* ether_src;
     const char* ether_dst;
+    const char* dst_addr;
 };
 
 extern "C"
diff --git a/src/protocols/ipv6.h b/src/protocols/ipv6.h
index 8003ad472..ecd965c68 100644
--- a/src/protocols/ipv6.h
+++ b/src/protocols/ipv6.h
@@ -29,6 +29,7 @@ namespace snort
 namespace ip
 {
 constexpr uint8_t IP6_HEADER_LEN = 40;
+constexpr uint8_t IP6_MAX_STR_LEN = 40; // 16 bytes in hexa = 32 + : delimiter every two bytes 7 + terminating 0 = 40
 constexpr uint32_t MIN_EXT_LEN = 8;
 constexpr uint8_t IP6_MULTICAST = 0xFF;  // first/most significant octet
 
